src/lasso/surge_2.rs

```diff
@@ -0,0 +1,660 @@
+      random_tape,
+    );
+
+    let proof_mem = CombinedTableEvalProof::prove(
+      &polynomials.combined_final_polys,
+      &eval_final,
+      &rand_mem,
+      &generators.final_commitment_gens,
+      transcript,
+      random_tape,
+    );
+
+    Self {
+      eval_dim,
+      eval_read,
+      eval_final,
+      proof_ops,
+      proof_mem,
+      eval_derefs,
+      proof_derefs,
+    }
+  }
+
+  fn verify<F1: Fn(usize) -> usize, F2: Fn(usize, &[<G>::ScalarField]) -> <G>::ScalarField>(
+    &self,
+    rand: (&Vec<<G>::ScalarField>, &Vec<<G>::ScalarField>),
+    grand_product_claims: &[super::gp_evals::GPEvals<<G>::ScalarField>], // NUM_MEMORIES-sized
+    memory_to_dimension_index: F1,
+    evaluate_memory_mle: F2,
+    commitments: &Self::Commitments,
+    generators: &Self::Generators,
+    r_hash: &<G>::ScalarField,
+    r_multiset_check: &<G>::ScalarField,
+    transcript: &mut merlin::Transcript,
+  ) -> Result<(), crate::utils::errors::ProofVerifyError> {
+    <Transcript as ProofTranscript<G>>::append_protocol_name(transcript, Self::protocol_name());
+
+    let (rand_mem, rand_ops) = rand;
+
+    // verify derefs at rand_ops
+    // E_i(r_i''') ?= v_{E_i}
+    self.proof_derefs.verify(
+      rand_ops,
+      &self.eval_derefs,
+      &generators.E_commitment_gens,
+      &commitments.E_commitment,
+      transcript,
+    )?;
+
+    let mut evals_ops: Vec<G::ScalarField> = Vec::new();
+    evals_ops.extend(self.eval_dim.clone());
+    evals_ops.extend(self.eval_read.clone());
+    evals_ops.resize(evals_ops.len().next_power_of_two(), G::ScalarField::zero());
+
+    // dim_i(r_i''') ?= v_i
+    // read_i(r_i''') ?= v_{read_i}
+    self.proof_ops.verify(
+      rand_ops,
+      &evals_ops,
+      &generators.dim_read_commitment_gens,
+      &commitments.dim_read_commitment,
+      transcript,
+    )?;
+
+    // final_i(r_i'') ?= v_{final_i}
+    self.proof_mem.verify(
+      rand_mem,
+      &self.eval_final,
+      &generators.final_commitment_gens,
+      &commitments.final_commitment,
+      transcript,
+    )?;
+
+    // verify the claims from the product layer
+    let init_addr = IdentityPolynomial::new(rand_mem.len()).evaluate(rand_mem);
+    for memory_index in 0..grand_product_claims.len() {
+      let dimension_index = memory_to_dimension_index(memory_index);
+      // Check ALPHA memories / lookup polys / grand products
+      // Only need 'C' indices / dimensions / read_timestamps / final_timestamps
+      Self::check_reed_solomon_fingerprints(
+        &grand_product_claims[memory_index],
+        &self.eval_derefs[memory_index],
+        &self.eval_dim[dimension_index],
+        &self.eval_read[dimension_index],
+        &self.eval_final[dimension_index],
+        &init_addr,
+        &evaluate_memory_mle(memory_index, rand_mem),
+        r_hash,
+        r_multiset_check,
+      )?;
+    }
+    Ok(())
+  }
+}
+
+impl<G: CurveGroup> SurgeFingerprintProof<G> {
+  fn check_reed_solomon_fingerprints(
```

Comment:
let's see if we can dedupe this logic with what's in `fingerprint_strategy.rs`

Reply:
TODO(JOLT-42)